File: web/src/lib/chips.ts
export interface ChipFilterDoc {
  slug: string;
  cuisine: string;
  course?: string;
  difficulty?: string;
  tags: string[];
  totalTime: number;
}

export type ChipKey =
  | "all"
  | "rapide"
  | "vege"
  | "asiatique"
  | "francais"
  | "dessert"
  | "favoris";

export interface FilterContext {
  favorites?: string[];
}

const CHIP_PREDICATES: Record<
  Exclude<ChipKey, "all" | "favoris">,
  (r: ChipFilterDoc) => boolean
> = {
  rapide: (r) => r.totalTime <= 30,
  vege: (r) => r.tags.includes("végé"),
  asiatique: (r) => r.tags.includes("asiatique"),
  francais: (r) => r.cuisine === "française",
  dessert: (r) => r.course === "dessert" || r.tags.includes("dessert"),
};

export function filterByChip(
  recipes: ChipFilterDoc[],
  chip: ChipKey,
  ctx: FilterContext = {},
): string[] {
  if (chip === "all") return recipes.map((r) => r.slug);
  if (chip === "favoris") {
    const favs = new Set(ctx.favorites ?? []);
    return recipes.filter((r) => favs.has(r.slug)).map((r) => r.slug);
  }
  const predicate = CHIP_PREDICATES[chip];
  return recipes.filter(predicate).map((r) => r.slug);
}

export type Course = "entrée" | "plat" | "dessert";

export function filterByCourse(
  recipes: ChipFilterDoc[],
  course: Course | null,
): string[] {
  if (!course) return recipes.map((r) => r.slug);
  return recipes.filter((r) => r.course === course).map((r) => r.slug);
}

export type Difficulty = "facile" | "moyenne" | "difficile";

export function filterByDifficulty(
  recipes: ChipFilterDoc[],
  difficulty: Difficulty | null,
): string[] {
  if (!difficulty) return recipes.map((r) => r.slug);
  return recipes
    .filter((r) => r.difficulty === difficulty)
    .map((r) => r.slug);
}

export function filterBySansGluten(
  recipes: ChipFilterDoc[],
  active: boolean,
): string[] {
  if (!active) return recipes.map((r) => r.slug);
  return recipes
    .filter((r) => r.tags.includes("sans gluten"))
    .map((r) => r.slug);
}

File: android/app/src/main/java/fr/vferries/cuisine/data/Chips.kt
package fr.vferries.cuisine.data

enum class ChipKey(val label: String) {
    ALL("Toutes"),
    FAVORIS("Favoris"),
    RAPIDE("Rapide"),
    VEGE("Végé"),
    ASIATIQUE("Asiatique"),
    FRANCAIS("Français"),
    DESSERT("Dessert"),
}

private val predicates: Map<ChipKey, (RecipeMeta) -> Boolean> = mapOf(
    ChipKey.ALL to { _ -> true },
    ChipKey.RAPIDE to { r -> r.totalTime <= 30 },
    ChipKey.VEGE to { r -> "végé" in r.tags },
    ChipKey.ASIATIQUE to { r -> "asiatique" in r.tags },
    ChipKey.FRANCAIS to { r -> r.cuisine == "française" },
    ChipKey.DESSERT to { r -> r.course == "dessert" || "dessert" in r.tags },
)

fun filterByChip(
    recipes: List<RecipeMeta>,
    chip: ChipKey,
    favorites: Set<String> = emptySet(),
): List<String> {
    if (chip == ChipKey.FAVORIS) {
        return recipes.filter { it.slug in favorites }.map { it.slug }
    }
    val predicate = predicates.getValue(chip)
    return recipes.filter(predicate).map { it.slug }
}

enum class Course(val label: String, val value: String) {
    ENTREE("Entrée", "entrée"),
    PLAT("Plat", "plat"),
    DESSERT("Dessert", "dessert");
}

fun filterByCourse(recipes: List<RecipeMeta>, course: Course?): List<String> {
    if (course == null) return recipes.map { it.slug }
    return recipes.filter { it.course == course.value }.map { it.slug }
}

enum class Difficulty(val label: String, val value: String) {
    FACILE("Facile", "facile"),
    MOYENNE("Moyenne", "moyenne"),
    DIFFICILE("Difficile", "difficile");
}

fun filterByDifficulty(
    recipes: List<RecipeMeta>,
    difficulty: Difficulty?,
): List<String> {
    if (difficulty == null) return recipes.map { it.slug }
    return recipes.filter { it.difficulty == difficulty.value }.map { it.slug }
}

fun filterBySansGluten(recipes: List<RecipeMeta>, active: Boolean): List<String> {
    if (!active) return recipes.map { it.slug }
    return recipes.filter { "sans gluten" in it.tags }.map { it.slug }
}
